flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerDriverTest.java

```diff
@@ -111,6 +111,32 @@ protected Context createContext() {
 		return new Context();
 	}
 
+	@Test
+	public void testRunAsyncCausesFatalError() throws Exception {
+		new Context() {{
+			final String exceptionMessage = "runAsyncCausesFatalError";
+			addContainerRequestFutures.add(new CompletableFuture<>());
+
+			testingYarnAMRMClientAsyncBuilder.setAddContainerRequestConsumer((ignored1, ignored2) ->
+					addContainerRequestFutures.get(addContainerRequestFuturesNumCompleted.getAndIncrement()).complete(null));
+			testingYarnAMRMClientAsyncBuilder.setGetMatchingRequestsFunction(ignored -> {
+				throw new RuntimeException(exceptionMessage);
+			});
+
+			final CompletableFuture<Throwable> throwableCompletableFuture = new CompletableFuture<>();
+			resourceEventHandlerBuilder.setOnErrorConsumer(throwableCompletableFuture::complete);
+
+			runTest(() -> {
+				runInMainThread(() -> getDriver().requestResource(testingTaskExecutorProcessSpec));
+				resourceManagerClientCallbackHandler.onContainersAllocated(ImmutableList.of(testingContainer));
+
+				Throwable t = throwableCompletableFuture.get(TIMEOUT_SEC, TimeUnit.SECONDS);
+				assertThat(ExceptionUtils.findThrowable(t, RuntimeException.class).isPresent(), is(true));
```

Comment:
`ExceptionUtils.findThrowable(t, RuntimeException.class)` could be pulled out to avoid the computation in the next line.